### PR TITLE
[mlir][emitc] Restrict integer and float types

### PR DESCRIPTION
**Pull request by TinaAMD**
Restrict which integers and floating-point types are valid in EmitC. This should cover the types which are supported in C++ and is aligned with what the emitter currently supports.

The checks are implemented as functions and not fully in tablegen to allow them to be re-used by conversions to EmitC.

### CODE REVIEW

File: mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td:25]:
```suggestion
def EmitCIntegerType : Type<CPred<"emitc::isSupportedIntegerType($_self)">,
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:55]:
```suggestion
def IntegerIndexOrOpaqueType : AnyTypeOf<[EmitCIntegerType, Index, EmitC_OpaqueType]>;
def FloatIntegerIndexOrOpaqueType : AnyTypeOf<[EmitCFloatType, IntegerIndexOrOpaqueType]>;
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td:28]:
```suggestion
def EmitCFloatType : Type<CPred<"emitc::isSupportedFloatType($_self)">,
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td:26]: Maybe
```suggestion
    "integer type supported by EmitC">;
```
?

File: mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td:29]: Maybe
```suggestion
    "floating-point type supported by EmitC">;
```
?

File: mlir/test/Dialect/EmitC/invalid_ops.mlir
Comment (marbre) [inline mlir/test/Dialect/EmitC/invalid_ops.mlir:173]:
```suggestion
    // expected-error @+1 {{'emitc.div' op operand #0 must be floating-point type supported by EmitC or integer type supported by EmitC or index or EmitC opaque type, but got 'tensor<i32>'}}
```

Comment (marbre) [reply mlir/test/Dialect/EmitC/invalid_ops.mlir:173]: Same for the tests below.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (marbre) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:57]:
```suggestion
bool mlir::emitc::isSupportedIntegerType(Type type) {
```

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (marbre) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:73]:
```suggestion
bool mlir::emitc::isSupportedFloatType(Type type) {
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.h
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.h:34]:
```suggestion
bool isSupportedIntegerType(mlir::Type type);
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.h
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.h:36]:
```suggestion
bool isSupportedFloatType(mlir::Type type);
```

**Review by marbre — CHANGES_REQUESTED** [review]
Thanks for working on this @TinaAMD! Some naming suggestions. Furthermore, could you please add some tests with invalid types? That would be really helpful.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td
Comment (TinaAMD) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td]: I renamed the types. While refactoring it, I noticed that the other types tend to follow the naming convention `EmitC_XType`. Should we go with `EmitC_IntegerType`/`EmitC_FloatType` maybe?

Comment (marbre) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td]: Well those types, to be precise `EmitC_ArrayType`, `EmitC_OpaqueType` and `EmitC_PointerType`, are EmitC specific and we have our own implementations for it. In contrast to that, `EmitCIntegerType` as wells as `EmitCFloatType`just refer to a collection of (bultin) types without implementing a new type, similar to

https://github.com/llvm/llvm-project/blob/4462bb58be7d68274fb9b443b7bf4803306df87d/mlir/include/mlir/Dialect/EmitC/IR/EmitC.td#L54-L55

Therefore, I am fine with the slightly differing name scheme. @simon-camp your thoughts?

Comment (simon-camp) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td]: I don't have a strong opinion either.

Comment (TinaAMD) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td]: Ok, sounds reasonable, let's go with `EmitCIntegerType`/`EmitCFloatType`.

File: mlir/test/Dialect/EmitC/invalid_types.mlir
Comment (TinaAMD) [inline mlir/test/Dialect/EmitC/invalid_types.mlir:99]: @marbre I added tests for invalid int and float types to the invalid type tests. Since we only have the checks enable for binary ops (so far), I used an operator in the test to show the error message. Was that the kind of test you would expect for the invalid types?

**Review by marbre — APPROVED** [review]
Thanks, lgtm! Let me know if I should merge on your behalf.

**Comment by TinaAMD** [comment]
I got commit access recently, so I should be able to merge, but thank you! I'll give @simon-camp some time to give feedback.

**Review by simon-camp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
